### ISSUE TITLE
Patch for High Sierra?

### ISSUE THREAD

**Issue by scottsanett** [Floris497/mac-pixel-clock-patch-V2#211]
I could be doing it wrong but the core display patch does not seem to be working for High Sierra. Will there be any update/fix for this?

**Comment by Floris497** [comment]
Could you try and force the latest version? (`CoreDisplay.command patch v2`)

**Comment by scottsanett** [comment]
That doesn't seem to work. The brightness change is noticeable but then I'm just stuck before the Apple logo splash screen appears.

**Comment by Floris497** [comment]
Hm, so you are not able to boot into Mac OS anymore?

**Comment by scottsanett** [comment]
Doesn't seem I can. I also deleted 0x12345678 from clover like I did with 10.12.

**Comment by Floris497** [comment]
Yep the patch has changed I can see.. I need to calculate a new one.. (also my mistake the last version is v3 not v2 but that does not change the difference that I need to make a v4) give me some time and I'll return with an update :)

**Comment by Floris497** [comment]
That was noticeable more difficult than previous times big changes in CoreDisplay and they removed almost all the naming in the executible.. (I thought I would not be able to find it..) I added a v4 to the script you can use it manually. please use with care.. because I have no idea if it will work as intended :)

**Comment by scottsanett** [comment]
I patched it v4 and I got "The code is still the same.. Patch did seem to run, but was probbably from the wrong version" again. I've posted the md5 just in case. 

     otool CoreDisplay: 961f97847f5234ea98cf97acb545be54
 otool CoreDisplay.bak: 961f97847f5234ea98cf97acb545be54
           CoreDisplay: 14f13ac79120596e4f0ac1a43fe8d4b1
       CoreDisplay.bak: b869a07a5efef0a1feff93c58fa476e6

**Comment by Floris497** [comment]
Uhg.. let me have a look when I get to my PC.. first some breakfast :)

**Comment by scottsanett** [comment]
I wonder if there is any tutorial/manual explaining the rationale of the perl script? I was just reading the instructions to patch later versions of Core Display from the original mac-pixel-clock-patch repo and tried it myself and that didn't work.

**Comment by Floris497** [comment]
Yes, exactly the IOKit thing is exactly the same.. but since now the code has changed a lot.. (maybe they are porting to Swift.. this removes most of the naming that was still available.. making it quite difficult to find the part of assembly that needs to be changed..) I will quickly test what I did wrong and come back to you.. sorry I kinda forgot to look at it..

**Comment by Floris497** [comment]
Also it seems I'm not on the latest version of 10.13.. I'm on beta 5 so I will update to 7 now

**Comment by Floris497** [comment]
Oh my bad.. I didn't fully enable v4 patch... and it looks like that the new patch does not work on beta 7.. let the search start again..

**Comment by Floris497** [comment]
It should now be working for 10.13 Beta 7

**Comment by scottsanett** [comment]
The code did seem to run. I diff'ed the two disassembled files and could see the changes. But after I cleared the field under ig-platform-id I was still stuck at the black screen.

**Comment by Floris497** [comment]
Wait but it does work without the patch?

**Comment by Floris497** [comment]
Btw, tell me what hardware you are running exactly and your platform id

**Comment by scottsanett** [comment]
Oops I was going to say I cleared the field under IntelGFX, not the ig-platform-id. I'm running Intel 6829hq with HD530 graphics. The platform-id I was using is 0x191b0000, but now I'm gonna give 0x19120000 a shot (which is what I use for 10.12) I will update as soon as I finish the test.

**Comment by Floris497** [comment]
Hm. I see, I have never played around with the ig-platform-id on my hack.. what time zone are you in?

**Comment by scottsanett** [comment]
So after I changed the platform id to 0x19120000 the black screen full of white characters was cleared and I was stuck at a fully black screen (with a slightly lighter hue) instead. 

> Does it work without the patch?
It doesn't seem so... I'm either stuck at a black screen with white characters or the scenario described above (due to with different platform ids).

I'm in China Standard.

**Comment by Floris497** [comment]
So it looks like your problem isn't my patch :) this is good then :) have other people attempted 10.13 with the same build? (And what do you see when you remote in to your computer when it's black screened?)

**Comment by scottsanett** [comment]
That seems to be the case. The patch is doing what it's supposed to do (redirecting the code to 0x1255a). I don't think many have tried installing high sierra on 4K laptops yet. Will try remoting later. This has been a lot harder than I expected.

**Comment by grigorovich** [comment]
MacOS 10.13 Beta (17A360a) . Public Beta 8 (not DB)
So. What "patch v#" I can to try ?

**Comment by scottsanett** [comment]
@grigorovich  Turns out the patch for HS is the same for Sierra. Try apply this manually in your terminal:
```
sudo perl -i.bak -pe 's|\xB8\x01\x00\x00\x00\xF6\xC1\x01\x0F\x85|\x33\xC0\x90\x90\x90\x90\x90\x90\x90\xE9|sg' /System/Library/Frameworks/CoreDisplay.framework/Versions/Current/CoreDisplay
sudo codesign -f -s - /System/Library/Frameworks/CoreDisplay.framework/Versions/Current/CoreDisplay
```
Beware, this patch breaks copy and paste and does not go well with iTunes.

**Comment by grigorovich** [comment]
What happened with copy/paste ?

**Comment by scottsanett** [comment]
Not sure if it is only me but i can't copy/paste after the patch is applied. iTunes also crashes whenever I open it.

**Comment by grigorovich** [comment]
@scottsanett Right. Me too. I tried vanila coredisplay.framework (and rebuilt cache) and copy/past works... So it could be help for somebody

**Comment by scottsanett** [comment]
@grigorovich but then the patch is gone isn't it?

**Comment by Floris497** [comment]
Well, there is a patch v4 that should work with the Beta's

**Comment by Floris497** [comment]
@scottsanett older patched can always be accessed by `patch.command patch vX` for example: 'CoreDisplay-patcher.command patch v3'

**Comment by grigorovich** [comment]
@scottsanett 2560x1080 display works in 1920x1080 right now. CoreDisplay not patched and i will wait for release bcoz "copy/paste" much more important for me than resolution

@Floris497 patch v3 works but "copy/paste" around all macos system stop work. I think @scottsanett caught that "feature" of new security method of "kernel/kext cache" specially for 10.13 ... Its just my opinion

**Comment by scottsanett** [comment]
@grigorovich Have you had any problem with iTunes with a patched CoreDisplay?

**Comment by grigorovich** [comment]
@scottsanett yes. But it crashed not always... Like randomly 50/50 and i just try to play only aac.. 
I have not patched coredisplay and all works stable right now (except resolution of second display)

**Comment by scottsanett** [comment]
@grigorovich iTunes crashes pretty much all the time here... But I do think the crashes and the patch are at least somewhat related.

**Comment by grigorovich** [comment]
In "Copy/Paste" problem i saw that in log:
process: pboard
`
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** initialization method -initWithFormat:locale:arguments: cannot be sent to an abstract object of class NSString: Create a concrete instance!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4cc170fb __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fff734c5c76 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff4cca8bfd +[NSException raise:format:] + 205
	3   Foundation                          0x0000000106605608 -[NSString initWithFormat:locale:arguments:] + 14
	4   Foundation                          0x00000001064181ad +[NSString stringWithFormat:] + 169
	5   Foundation                          0x0000000106433827 -[NSArray(NSArray) encodeWithCoder:] + 816
	6   Foundation                          0x00007fff4ec0493e -[NSXPCEncoder _encodeObject:] + 578
	7   UserActivity                        0x00007fff701d349d -[UASharedPasteboard<…>`

and then plenty apps (WiFiAgent, Finder, AppleShell,Safari... 20+) reports about pboard like :
`
Connection to 'pboard' server had an error: <error: 0x7fffad7b6c40> { count = 1, transaction: 0, voucher = 0x0, contents =
	"XPCErrorDescription" => <string: 0x7fffad7b6fb0> { length = 22, contents = "Connection interrupted" }
}`

So i think its common problem outside of 'pixel clock patch' ...

**Comment by scottsanett** [comment]
@grigorovich But pboard doesn't crash without the patch, no?

**Comment by grigorovich** [comment]
@scottsanett doesn't. i think its kernel/security issue bcoz pboard works fine on macbook with macos PB8 w/o patched coredisplay.

**Comment by scottsanett** [comment]
@grigorovich I replaced the patch with CoreDisplayFixup.kext and Lilu.kext (and unpatched CoreDisplay). Copy and paste now works. iTunes still crashes though.

**Comment by gothickitty93** [comment]
@Floris497 CoreDisplay for High Sierra GM should have accidentally found it's way to your email.

**Comment by Floris497** [comment]
I'll start testing tomorrow.. I just finished another project :)

**Comment by TylerD89** [comment]
Hey man can you ass the md5 hash:
---- BEGINNING MD5 HASH SUMS ---- version: 10.13 build:17A365

     otool CoreDisplay: 73e3dfcd3df91cb8df6c0c871a4eaf19
 otool CoreDisplay.bak: 73e3dfcd3df91cb8df6c0c871a4eaf19
           CoreDisplay: 23aef373352d6bb5406f5fd6769929f8
       CoreDisplay.bak: 92d03aef881f9f9de05298ac07d92670

---- ENDING MD5 HASH SUMS -------

Thanks

**Comment by TheTrollCaptain** [comment]
Well crap... I finally got patchv4 to run and it said that it was applied successfully. However, macOS doesn't boot anymore... Sheet. How can I manually restore the CoreDiplay backups through the filesystem (when I plug my SSD into my other Mac)?

**Comment by Floris497** [comment]
Ehm.. Patch 4 isn't ready.. it's at least breaking copy paste.. some fix might be in the issues but I had no time figuring it out already

**Comment by cbjerg** [comment]
Tried v4 and it broke the core display. It didn’t create a backup. Anybody want to send me CoreDisplay from fresh High Sierra

**Comment by Floris497** [comment]
There is a backup in the same location replace it with `mv file.bak file`

**Comment by cbjerg** [comment]
There was no Bak file created. I am looking for it now in the CoreDisplay.framework/Versions/A/ location

**Comment by Floris497** [comment]
Hm.. let me check.. (you can just reinstall macOS, probably fastest way to fix it)

**Comment by cbjerg** [comment]
There was a bak file. Just not visible to "ls -la"   I ran the command and got back in. Thank you. Waiting for patch to be updated then

**Comment by TheTrollCaptain** [comment]
Ahh okay. Thanks man

**Comment by TheTrollCaptain** [comment]
@cjberg It's easier to make a copy of the framework, change the extension to .zip, and access it that way. I found terminal the "links" inside of the framework make it very confusing to find the files

**Comment by TheTrollCaptain** [comment]
Crap.. Well my copy/paste is broken... Will the future patch sort it out or do I have to reinstall the entire OS

**Comment by Floris497** [comment]
Hm, you can run the unpatch command.. for me myself it seems to break everything.. doesn't recognize my internal screen anymore. I was looking around a little bit how to fix it.. but I have little time..

**Comment by TheTrollCaptain** [comment]
I'll try that thanks. Thanks for your help

**Comment by TheTrollCaptain** [comment]
I just unpatched it but it still is broken... High Sierra breaks everything :(

**Comment by Floris497** [comment]
Unfortunately I have had the same happen to me in the beta.. no clue why :/

**Comment by TheTrollCaptain** [comment]
It may just be a bug or something... I really don’t feel like backing up and reinstalling so I’ll just see if it gets fixed in the future

**Comment by Floris497** [comment]
Hm.. i had to reinstall to get it working again.. if you ask me no backup is really needed.. just run the installer again :) (don't blame me if you did need a backup in the end :p)

**Comment by grigorovich** [comment]
@scottsanett 
for fix resolutions and solve copy/paste at same time i apply patch: 
[toleda/intel_graphics](https://github.com/toleda/graphics_Intel_framebuffers/tree/master/HD4600%2B-Azul%20Edit)
Before apply just choose your connector configuration. For me is patch with '_c' letter.

and set 2 type of resolutions 2560x1080 in SwitchResX (50Hz, 60Hz)

Now second display with 2560x1080 works fine and copy/paste works as well.
MacOS 10.13.1 dev.

Please report if it helps

**Comment by TheTrollCaptain** [comment]
Sweet! I'll try this when I get home for lunch. What are your specs?

**Comment by erbrango** [comment]
Hello. Have iMac 4K end 2015 and get 60hz with external 4K TV not. Is this at all feasible?

**Comment by TheTrollCaptain** [comment]
@erbrango Wait so you want 60Hz on the 4K TV with the patch? What is the TV running at without the patch? (eg 1080p @ 30Hz)

**Comment by erbrango** [comment]
Mit Capitan geht 60hz ohne nichts und ab Sierra / High Sierra geht überhaupt nicht

**Comment by shad00m** [comment]
I tried the pixelclockpatch and the lilo+corevideo kext posted here
but I cant push above 330Mhz, Is the perl script above confirmed working to push past the 330mhz limit?

**Comment by jb107502** [comment]
Any luck on a new official version of CoreDisplay-patcher? I am currently using 10.13.1. I am extremely new to any coding and really only use terminal for this program and getting full use of my ultra wide monitor.

**Comment by TylerD89** [comment]
@shad00m Since display refresh rate is measured in Hz and the highest refresh rate on the market is 144Hz, it is quite impossible to get 300MHz (1000000Hz = 1 MHz)... ;)

**Comment by TylerD89** [comment]
@jb107502 Nope. The Copy + Paste issue still exists... CoreDisplayPatcher + Lilu is currently the only way to get full resolution + refresh support

**Comment by TylerD89** [comment]
Or I guess you could try SwitchX but it does require more tampering

**Comment by jb107502** [comment]
@TylerD89 I have SwitchX. It being meh led me to Floris.

**Comment by TylerD89** [comment]
@jb107502 Yeah I’ve heard it’s terrible

**Comment by TahaSolutions** [comment]
I also need the patch for 10.13.1 

Appreciate to get notified when it's ready without breaking anything :)

**Comment by daniel-hdez-lopez** [comment]
Hi,

 For those of you that went the lilu+coredisplay way and don't have copy/paste functionality, try executing 'sudo update_dyld_shared_cache'. I regained the functionality after executing it. 

Cheers.

**Comment by Floris497** [comment]
Would that work with the patch as well?? I'll try when home

**Comment by daniel-hdez-lopez** [comment]
I don't know. I installed lilu+coredisplay kexts because this patch doesn't work for me at the moment.

**Comment by TylerD89** [comment]
Interesting, how the hell did you come up with that lol

**Comment by daniel-hdez-lopez** [comment]
Pure luck. Playing with the dlyd cache is something you try when something's off with your mac as infinite wheel spinning on some applications among other faults.... so i tried the same that worked for me before. This happens on original macs too... it's annoying.

**Comment by TylerD89** [comment]
Yeah I don’t even know what the DLYD cache is so props dude

**Comment by Floris497** [comment]
@daniel-hdez-lopez seems to work nicely.. it was mentioned before somewhere in issues but I forgot about it. anyway I made a whole new patch, because the old one was really broken.. expect a new patch soon (only need a bit of testing)

**Comment by jb107502** [comment]
Now THIS is my favorite thing I got for Xmas. I am a photographer and not having the Ultrawide use of my monitor was killing me. Thanks so much everyone and @Floris497!

**Comment by Floris497** [comment]
Enjoy! Merry belated Christmas! And a happy new year!

**Comment by TahaSolutions** [comment]
Is this fix final now?

Any step by step guide to install this patch?

Br
Shahrouz

**Comment by Floris497** [comment]
It is.. no step by step.. there is some info in readme should ne sufficient

**Comment by TahaSolutions** [comment]
Thanks a lot.

Just managed to install the patch. 

Wish you and all the guys here a very successful year in 2018.

Regards
Shahrouz Raeisi

**Comment by aShure** [comment]
v4 patch doesn't seem to work for a MBP 2017 (Radeon Pro 555).

A matching md5 sum was not found, so I forced v4.

**Comment by TylerD89** [comment]
@aShure What resolution are you trying to get on your MBP?

**Comment by TylerD89** [comment]
@aShure and also, be sure you have the latest version of the patch. V4 is out so you shouldn’t have to force it...

**Comment by aShure** [comment]
@TylerD89 had to force it as the md5 sum didn't match. Latest High Sierra up to date, latest commit from git. Trying to use UHD 2160p60. USB-C to Displayport adapter.
Screen, adapter, cable runs fine with 60Hz on Windows hosts.

**Comment by Floris497** [comment]
Did it work after forcing? Could you post MD5 output? :)

**Comment by aShure** [comment]
Sadly, it didn't work after forcing. See #242 for the MD5 output, I posted a log of the output when I first tried to force it.

**Comment by Floris497** [comment]
Update to 10.13.2 :)

**Comment by aShure** [comment]
Great idea :) 
It seems that the usual path of updating via the app store is not available with 10.13.1 -> 10.13.2. It even says so on https://support.apple.com/kb/DL1946?locale=en_US 
Apple, why oh why?

**Comment by Floris497** [comment]
Oh?? Really? That's strange..?

**Comment by aShure** [comment]
My bad, it just didn't show up in the app store as it was downloading in the background. Still a horrible UX :(

I patched CoreDisplay with autodetected correct MD5 now:
`(otool) Detected patched CoreDisplay on OS X 10.13.2 17C88.`

Alas, SwitchResX still won't let me set 60Hz:

![image](https://user-images.githubusercontent.com/4117408/34443065-c558064e-ecc6-11e7-8321-a55d14411b84.png)

**Comment by dragzter** [comment]
I have macbook pro 2018 and im getting the unkown coredispay patch message.  My specs

Radeon Pro 555X 4096 MB
Intel UHD Graphics 630 1536 MB

My output.

SIP looks to be disabled, all good!

Unknown version of CoreDisplay found..
---- BEGINNING MD5 HASH SUMS ---- version: 10.13.6 build:17G2307

     otool CoreDisplay: 20f0d2491e9370d3d74d25740aa855ba
 otool CoreDisplay.bak: NO FILE (this is okay)
           CoreDisplay: 24349340f35fb931216e0ccdab2ad3c1
       CoreDisplay.bak: NO FILE (this is okay)

---- ENDING MD5 HASH SUMS -------

any idea how i can get this to work?  I have a 4k lg monitor. 

LG Ultra HD:
  Resolution:	3840 x 2160 (2160p 4K UHD - Ultra High Definition)
  UI Looks like:	1920 x 1080 @ 30 Hz
  Framebuffer Depth:	30-Bit Color (ARGB2101010)
  Mirror:	Off
  Online:	Yes
  Rotation:	Supported
  Automatically Adjust Brightness:	No
  Connection Type:	DVI or HDMI
  Adapter Firmware Version:	7.55

I cant get it to budge from 30hz

**Comment by dragzter** [comment]
I tried patching with v4 and v3

**Comment by dragzter** [comment]
My OS is High Sierra 10.13.6
